File: C++/Flatten Binary Tree to Linked List.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if (root == NULL) return;
        if (root->left != NULL) {
        	TreeNode *leftNode = root->left;
        	TreeNode *rightNode = root->right;
        	root->left = NULL;
        	root->right = leftNode;
        	TreeNode* p = leftNode;
        	while (p->right != NULL) {
        		p = p->right;
        	}
        	p->right = rightNode;
        }
        flatten(root->right);
    }
};

File: Python/Flatten Binary Tree to Linked List.py
#!/usr/bin/env python

from collections import deque
import collections

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# @param root, a tree node
# @return nothing, do it in place

def preorder(self, root, q):
    if root == None:
        return q
    q.append(root)
    q = self.preorder(root.left, q)
    q = self.preorder(root.right, q)
    return q

def flatten(self, root):
    q = []
    self.preorder(root, q)
    if len(q) == 0:
        return None
    root = q.pop(0)
    head = root
    root.left = None
    while len(q) > 0:
        root.right = q.pop(0)
        root = root.right
        root.left = None
    root.right = None
    return head